[#5962] feat(client): added audit cli command model

VigneshSK17 opened the pull request:
### What changes were proposed in this pull request?

The audit command is one of the commands suggested by @justinmclean as part of adding Model entity support for the CLI.

Also includes addition of relevant testing code for Model entity CLI commands a whole.

### Why are the changes needed?

To add audit functionality for a Model using the CLI

Improvement: #5962 

### Does this PR introduce _any_ user-facing change?

Yes.

The audit command for a model was added.

### How was this patch tested?

Unit tests were added for Model CLI support and ran successfully for the audit command.

VigneshSK17 commented:
@justinmclean Could you please review this?

I was a bit unclear on what the ModelAudit command should be auditing, please let me know if there's more functionality I've yet to add.

Thank you.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/TestableCommandLine.java:
```diff
@@ -21,118 +21,8 @@
-import org.apache.gravitino.cli.commands.UserAudit;
-import org.apache.gravitino.cli.commands.UserDetails;
+
+import org.apache.gravitino.cli.commands.*;
```
We use Spotless, and this is likely to break the build, please don't use wildcards.

tengqm replied:
@justinmclean Can you elaborate why this revised import is not an improvement?

justinmclean replied:
We don't use wildcard imports, and that is set up in a rule in Spotless. You'll need to ask @jerryshao why.

tengqm replied:
I just found the customized rule here:
https://github.com/apache/gravitino/blob/main/build.gradle.kts#L140-L144

jerryshao replied:
Please check this (https://errorprone.info/bugpattern/WildcardImport#:~:text=Wildcard%20imports%20are%20forbidden%20by,for%20programmers%20and%20for%20tools.).

VigneshSK17 replied:
Makes sense, seems like that got automated somehow. Fixed.

tengqm replied:
Thanks for the pointer, @jerryshao. I do agree that for each and every rule in a lint tool, there is a very clear and solid reason. Java is no exception.
However, I don't think that rule applies universally to all cases.
On the contrary, my gut feeling is that the specific scenario this rule is protecting is a trivial one. It may not even apply to Gravatino.
The fact that this rule has to be EXPLICITLY ENABLED in the gradle script may be a hint for us. It is not something mandatory for most projects.

The stackoverflow discussion has some valid points as well, despite the disputes aroused. `import java.awt.*` is cumbersome and stupid, because it may introduce a lot of name conflicts IF you are overriding a lot of the existing widgets. However, import all exception classes under my own package `very.specific.package.to.my.project.exceptions.*` should be a good practice.
We know what we are doing. We create rules and abide by them when it improves code quality and/or productivity. We ignore/abandon rules when they are counterproductive and/or harmful to the code base.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoCommandLine.java:
```diff
@@ -494,6 +496,42 @@ private void handleTableCommand() {
+    if (metalake == null) missingEntities.add(CommandEntities.METALAKE);
+    if (catalog == null) missingEntities.add(CommandEntities.CATALOG);
+
+    // TODO: Add remaining commands to TestableCommandLine
```
No need for this TODO.

VigneshSK17 replied:
Fixed

justinmclean commented:
Looks good, but you need to update the help text and CLI documentation.

tengqm commented on clients/cli/src/main/java/org/apache/gravitino/cli/GravitinoCommandLine.java:
```diff
@@ -494,6 +496,42 @@ private void handleTableCommand() {
+    // TODO: Add remaining commands to TestableCommandLine
+    String model = name.getModelName();
+    if (model == null) {
+      missingEntities.add(CommandEntities.MODEL);
```
This is getting complicated ... unnecessarily.
Is it possible to postpone these argument checking to the specific command action handlers? For example, we only check the arguments in, say, `newModelAudit` layer.
It is hard, if possible at all, to do a unified check at this layer.
For example, you may want to list models, in which case the model name would be empty.

justinmclean replied:
There is a separate issue on adding a validate method to commands, but until that is done this is OK.

tengqm replied:
Agreed. Let's keep the scope of this PR.

VigneshSK17 commented:
> Looks good, but you need to update the help text and CLI documentation.

ahh knew I forgot something. Will add shortly.

VigneshSK17 commented:
Added `clients/cli/src/main/resources/model_help.txt` and updated `docs/cli.md` with Model. Let me know if any other documentation needs to be updated.

Also happy new year everyone.

justinmclean commented:
The documentation looks good, however, the CI is failing due to a license header issue, which needs to be fixed.

VigneshSK17 commented:
Hopefully everything is fixed now. Thank you for mentioning that.

tengqm commented on clients/cli/src/main/resources/model_help.txt:
```diff
@@ -0,0 +1,8 @@
+Example commands
+
+Show model audit information
+gcli model details --name catalog_postgres.hr --audit
```
This text doesn't seem to have an explicit format.
I know where we come from, but this can be improved.

justinmclean replied:
I'm not sure what you mean by "explicit format" in this context. This follows the format of the other help files.

tengqm commented on docs/cli.md:
````diff
@@ -27,7 +27,7 @@ Or you use the `gcli.sh` script found in the `clients/cli/bin/` directory to run
 
  ```bash
-  usage: gcli [metalake|catalog|schema|table|column|user|group|tag|topic|fileset] [list|details|create|delete|update|set|remove|properties|revoke|grant] [options]
+  usage: gcli [metalake|catalog|schema|model|table|column|user|group|tag|topic|fileset] [list|details|create|delete|update|set|remove|properties|revoke|grant] [options]
````
how about `usage: gcli <resource> <action> [options]` ?

justinmclean replied:
Because that only gives a vague suggestion to the user, it's more helpful for them to see the possible entities and actions.

VigneshSK17 commented:
Everything should be accounted for this time. Apologies for the constant fixes, just getting up to speed :)

justinmclean reviewed (approved):
Thanks for your contribution.
